realm/src/main/java/io/realm/Realm.java

```diff
@@ -1873,4 +1876,44 @@ public static Object getDefaultModule() {
         void execute(Realm realm);
     }
 
+    /**
+     * Encapsulates an async {@link RealmQuery}.
+     * <p>
+     * This will run the {@link RealmQuery} on a worker thread, then invoke this callback on the caller thread
+     */
+    public interface QueryCallback<E extends RealmObject> {
+        void onSuccess (RealmResults<E>  results);
+        void onError (Throwable t);
+    }
+
+    /**
+     * Used for debugging/testing purpose to add any logic (within the caller's thread)
+     * before we return the results
+     */
+    public interface DebugQueryCallback<E extends RealmObject> extends QueryCallback<E> {
+        /**
+         * Runs on the caller's thread just before we hand over the result to {@link #onSuccess(RealmResults)}
+         */
+        void onBackgroundQueryCompleted(Realm realm);
+    }
+
+    //FIXME Realm.java being the public API and the implementation.
```

Comment:
Yeah, we probably need to discuss how this should be done. Some other requirements to consider:

- Support DynamicRealm
- Supporting JavaRealm
- Supporting AndroidRealm